src/placement/mapless_jump_map.c

```diff
@@ -0,0 +1,1404 @@
+ *			the root.
+ */
+uint64_t get_dom_cnt(struct pool_domain *dom)
+{
+	uint64_t count = 0;
+
+	if (dom->do_children != NULL) {
+		int i;
+
+		count += dom->do_child_nr;
+
+		for (i = 0; i < dom->do_child_nr; ++i)
+			count += get_dom_cnt(&(dom->do_children[i]));
+	}
+
+	return count;
+}
+
+/**
+ * Jump Consistent Hash Algorithm that provides a bucket location
+ * for the given key. This algorithm hashes a minimal (1/n) number
+ * of keys to a new bucket when extending the number of buckets.
+ *
+ * \param[in]   key             A unique key representing the object that
+ *                              will be placed in the bucket.
+ * \param[in]   num_buckets     The total number of buckets the hashing
+ *                              algorithm can choose from.
+ *
+ * \return                      Returns an index ranging from 0 to
+ *                              num_buckets representing the bucket
+ *                              the given key hashes to.
+ */
+static inline uint32_t
+jump_consistent_hash(uint64_t key, uint32_t num_buckets)
+{
+	int64_t z = -1;
+	int64_t y = 0;
+
+	while (y < num_buckets) {
+		z = y;
+		key = key * 2862933555777941757ULL + 1;
+		y = (z + 1) * ((double)(1LL << 31) /
+			((double)((key >> 33) + 1)));
+	}
+	return z;
+}
+
+/**
+ * Assembly instruction for fast 4-byte CRC computation on Intel X86
+ * Copied from BSD-licensed Data Plane Development Kit (DPDK) rte_hash_crc.h
+ *
+ * \param[in] data              Primary input to hash function
+ * \param[in] init_value        Acts like a seed to the CRC, useful to get
+ *                              different results with the same data
+ */
+static inline uint32_t
+crc32c_sse42_u32(uint32_t data, uint32_t init_val)
+{
+	__asm__ volatile(
+		"crc32l %[data], %[init_val];"
+		: [init_val] "+r"(init_val)
+		: [data] "rm"(data));
+	return init_val;
+}
+
+/**
+ * Computes an 8-byte CRC using fast assembly instructions.
+ *
+ * This is useful for mapless placement to pseudorandomly permute input keys
+ * that are similar to each other. This dramatically improves the even-ness of
+ * the distribution of output placements.
+ *
+ * Since CRC is typically a 4-byte operation, this by computes a 4-byte CRC of
+ * each half of the input data and concatenates them
+ *
+ * \param[in] data              Primary input to hash function
+ * \param[in] init_value        Acts like a seed to the CRC, useful to get
+ *                              different results with the same data
+ */
+static inline uint64_t
+crc(uint64_t data, uint32_t init_val)
+{
+	return (uint64_t)crc32c_sse42_u32((data & 0xFFFFFFFF), init_val)
+		| ((uint64_t)crc32c_sse42_u32(((data >> 32) & 0xFFFFFFFF),
+					init_val) << 32);
+}
+
+/** add one failed shard into remap list */
+static void
+remap_add_one(d_list_t *remap_list, struct failed_shard *f_new)
+{
+	struct failed_shard 	*f_shard;
+	d_list_t			*tmp;
+	D_DEBUG(DB_PL,"fnew: %u",f_new->fs_shard_idx);
```

Comment:
(style)  Missing a blank line after declarations
(style)  space required after that ',' (ctx:VxV)
(style)  space required after that ',' (ctx:VxV)